src/binder/bind/bind_table_function.cpp

```diff
@@ -41,11 +42,25 @@ BoundTableScanInfo Binder::bindTableFunc(const std::string& tableFuncName,
         entry->ptrCast<catalog::FunctionCatalogEntry>());
     validateParameterType(positionalParams);
     auto tableFunc = func->constPtrCast<TableFunction>();
+    std::vector<common::LogicalType> inputTypes;
+    if (tableFunc->inferInputTypes) {
+        // For functions which take in nested data types, we have to use the input parameters to
+        // detect the input types. (E.g. query_hnsw_index takes in an ARRAY which needs the user
+        // input parameters to decide the array dimension).
+        inputTypes = tableFunc->inferInputTypes(positionalParams);
```

Comment:
interesting. I would comment two things in the future.

It make more sense to have vector index template implementation of all possible types. Otherwise we end with different design principles.

If in the future we indeed plan to remove this inferInputTypes functor. Then in this PR I would hide the change to vector index only, i.e. remove the functor and let vector index fold in its own bind function.

I'll leave the decision to u.

If you want to keep inferInputTypes, it's better to call it forceInputTypes